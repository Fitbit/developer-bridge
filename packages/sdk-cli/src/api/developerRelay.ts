import * as t from 'io-ts';
import { decode } from '@fitbit/jsonrpc-ts';
import stream from 'stream';
import websocketStream from 'websocket-stream';

import { apiFetch } from './baseAPI';
import { assertContentType, assertOK, assertJSON, decodeJSON } from '../util/fetchUtil';

// tslint:disable-next-line:variable-name
export const Host = t.type(
  {
    id: t.string,
    displayName: t.string,
    roles: t.array(t.string),
    state: t.union([
      t.literal('available'),
      t.literal('busy'),
    ]),
  },
  'Host',
);
export type Host = t.TypeOf<typeof Host>;

// tslint:disable-next-line:variable-name
const HostsResponse = t.type(
  {
    hosts: t.array(Host),
  },
  'HostsResponse',
);

// tslint:disable-next-line:variable-name
const Hosts403Response = t.type(
  {
    errors: t.array(t.type({
      message: t.string,
    })),
  },
  'Hosts403Response',
);
type Hosts403Response = t.TypeOf<typeof Hosts403Response>;

async function getConnectionURL(hostID: string) {
  const response = await apiFetch(
    `1/user/-/developer-relay/hosts/${hostID}`,
    { method: 'POST' },
  ).then(assertOK).then(assertContentType('text/uri-list'));
  const uriList = (await response.text())
    .split('\r\n')
    .filter(line => line[0] !== '#');
  return uriList[0];
}

function createWebSocket(uri: string) {
  return new Promise<stream.Duplex>((resolve, reject) => {
    const ws = websocketStream(uri, { objectMode: true });
    ws.on('connect', () => resolve(ws));
    ws.on('error', e => reject(e));
  });
}

export async function connect(hostID: string) {
  const url = await getConnectionURL(hostID);
  return createWebSocket(url);
}

export async function hosts() {
  const response = await apiFetch('1/user/-/developer-relay/hosts.json')
    .then(async (response: Response) => {
      if (response.ok || response.status !== 403) return response;

      const errorObj = await assertJSON()(response).then(decode(Hosts403Response));
      const errorString = errorObj.errors.join('\n');
      throw new Error(errorString);
    })
    .then(decodeJSON(HostsResponse));

  const hostsWithRole = (role: string) =>
    response.hosts.filter(host => host.roles.includes(role));

  return {
    appHost: hostsWithRole('APP_HOST'),
    companionHost: hostsWithRole('COMPANION_HOST'),
  };
}
