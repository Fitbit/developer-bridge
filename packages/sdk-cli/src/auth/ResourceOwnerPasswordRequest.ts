export interface ResoureOwnerTokenRequestJson {
  grant_type: 'password';
  client_id: string;
  username: string;
  password: string;
  extras?: Record<string, string>;
}

export default class ResoureOwnerTokenRequest {
  clientId: string;
  username: string;
  password: string;
  extras?: Record<string, string>;

  constructor(request: ResoureOwnerTokenRequestJson) {
    this.clientId = request.client_id;
    this.username = request.username;
    this.password = request.password;
    this.extras = request.extras;
  }

  /**
   * Serializes a TokenRequest to a JavaScript object.
   */
  toJson(): ResoureOwnerTokenRequestJson {
    return {
      grant_type: 'password',
      username: this.username,
      password: this.password,
      client_id: this.clientId,
      extras: this.extras,
    };
  }

  toStringMap(): Record<string, string> {
    const map: Record<string, string> = {
      grant_type: 'password',
      client_id: this.clientId,
      username: this.username,
      password: this.password,
    };

    // copy over extras
    if (this.extras) {
      for (const extra in this.extras) {
        if (this.extras.hasOwnProperty(extra) && !map.hasOwnProperty(extra)) {
          // check before inserting to requestMap
          map[extra] = this.extras[extra];
        }
      }
    }
    return map;
  }
}
