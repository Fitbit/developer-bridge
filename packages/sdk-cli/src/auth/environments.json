{
  "production": {
    "clientId": "228TQF",
    "apiUrl": "https://api.fitbit.com:443",
    "devRelayUrl": "wss://027-v3-api-soa.fitbit.com:443/dbridge/"
  },
  "stage": {
    "clientId": "228VJK",
    "apiUrl": "https://api.fitbit-stage.com:443",
    "devRelayUrl": "wss://developer-relay-stage-frontend.site-ops.fitbit.com/"
  },
  "int": {
    "clientId": "228W5N",
    "apiUrl": "https://api.fitbit-int.com:443",
    "devRelayUrl": "wss://developer-relay-int-frontend.site-ops.fitbit.com/"
  }
}
