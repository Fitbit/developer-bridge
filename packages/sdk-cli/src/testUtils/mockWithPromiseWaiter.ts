export default function mockWithPromiseWaiter<T>(
  mockFn: jest.Mock,
  returnValue?: any,
) {
  return new Promise<T>((resolve) => {
    mockFn.mockReset();
    mockFn.mockImplementationOnce((arg: T) => {
      resolve(arg);
      return returnValue;
    });
  });
}
