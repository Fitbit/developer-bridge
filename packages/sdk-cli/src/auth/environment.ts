import process from 'process';
// tslint:disable-next-line:import-name
import environments from './environments.json';

export default () => {
  const defaultEnv = 'production';
  const environment = (
    process.env.FITBIT_SDK_ENVIRONMENT || defaultEnv
  ) as keyof typeof environments;

  if (Object.keys(environments).indexOf(environment) === -1) {
    throw new Error(`Invalid environment specified: ${environment}`);
  }

  const config = environments[environment];

  const clientId = process.env.FITBIT_SDK_CLIENT_ID;

  if (clientId) {
    config.clientId = clientId;
  }

  return {
    environment,
    config,
  };
};
