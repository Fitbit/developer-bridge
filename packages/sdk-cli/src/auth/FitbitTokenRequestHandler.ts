import {
  AuthorizationServiceConfiguration,
} from '@openid/appauth/built/authorization_service_configuration';
import { AppAuthError } from '@openid/appauth/built/errors';
import { RevokeTokenRequest } from '@openid/appauth/built/revoke_token_request';
import {
  ErrorType,
  TokenError,
  TokenResponse,
} from '@openid/appauth/built/token_response';
import {
  TokenRequestHandler,
} from '@openid/appauth/built/token_request_handler';
import { BasicQueryStringUtils } from '@openid/appauth/built/query_string_utils';
import * as t from 'io-ts';

import environment from './environment';
import fetch from '../fetch';

const { clientId, clientSecret } = environment().config;

// Fitbit's error responses are non-standard per
// https://tools.ietf.org/html/rfc6749#section-5.2
// tslint:disable-next-line:variable-name
const FitbitAuthErrorResponse = t.interface({
  errors: t.array(
    t.interface({
      errorType: t.string,
      message: t.string,
    }),
  ),
  success: t.boolean,
});
type FitbitAuthErrorResponse = t.TypeOf<typeof FitbitAuthErrorResponse>;

// tslint:disable-next-line:variable-name
const FitbitTokenResponse = t.interface({
  access_token: t.string,
  refresh_token: t.string,
  expires_in: t.number,
});
type FitbitTokenResponse = t.TypeOf<typeof FitbitTokenResponse>;

const commonParams = {
  method: 'POST',
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded',
    Authorization: 'Basic ' + Buffer.from(`${clientId}:${clientSecret}`).toString('base64'),
  },
};

export default class FitbitTokenRequestHandler implements TokenRequestHandler {
  public readonly utils = new BasicQueryStringUtils();

  performRevokeTokenRequest(
    configuration: AuthorizationServiceConfiguration,
    request: RevokeTokenRequest,
  ): Promise<boolean> {
    return fetch(
      configuration.revocationEndpoint,
      {
        ...commonParams,
        body: this.utils.stringify({
          token: request.token,
        }),
      },
    ).then(response => response.ok);
  }

  async performTokenRequest(
    configuration: AuthorizationServiceConfiguration,
    request: {
      toStringMap: () => Record<string, string>,
    },
  ): Promise<TokenResponse> {
    const response = await fetch(
      configuration.tokenEndpoint,
      {
        ...commonParams,
        body: this.utils.stringify(request.toStringMap()),
      },
    );
    const responseJson = await response.json();

    if (response.status === 200 && FitbitTokenResponse.is(responseJson)) {
      return new TokenResponse(responseJson);
    }

    if (FitbitAuthErrorResponse.is(responseJson)) {
      const error = responseJson.errors[0];
      throw new AppAuthError(
        error.errorType,
        new TokenError({
          error: error.errorType as ErrorType,
          error_description: error.message,
        }),
      );
    }

    throw new AppAuthError(
      `Unexpected response format for status ${response.status}: ${JSON.stringify(responseJson)}`,
    );
  }
}
