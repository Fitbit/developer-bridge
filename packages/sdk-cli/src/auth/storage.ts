import { TokenResponse } from '@openid/appauth/built/token_response';
import * as t from 'io-ts';
import keytar from 'keytar';

import environment from './environment';

const keyStoreService = 'fitbit-sdk';
const keyStoreAccount = environment().environment;

// tslint:disable-next-line:variable-name
const AuthStorage = t.interface({
  access_token: t.string,
  refresh_token: t.string,
  issued_at: t.number,
  expires_in: t.string,
});
type AuthStorage = t.TypeOf<typeof AuthStorage>;

function set(tokenResponse: TokenResponse) {
  return keytar.setPassword(
    keyStoreService,
    keyStoreAccount,
    JSON.stringify(tokenResponse.toJson()),
  );
}

async function get(): Promise<TokenResponse | null> {
  const authDataStr = await keytar.getPassword(keyStoreService, keyStoreAccount);
  if (authDataStr === null) return null;
  let authData;
  try {
    authData = JSON.parse(authDataStr);
  } catch (ex) {
    await clear();
    return null;
  }
  if (!AuthStorage.is(authData)) {
    await clear();
    return null;
  }
  return new TokenResponse(authData);
}

function clear() {
  return keytar.deletePassword(keyStoreService, keyStoreAccount);
}

export default {
  set,
  get,
  clear,
};
