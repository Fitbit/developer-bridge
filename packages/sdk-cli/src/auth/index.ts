import { AuthorizationRequest } from '@openid/appauth/built/authorization_request';
import {
  AuthorizationNotifier,
  AuthorizationRequestHandler,
} from '@openid/appauth/built/authorization_request_handler';
import { AuthorizationResponse } from '@openid/appauth/built/authorization_response';
import {
  AuthorizationServiceConfiguration,
} from '@openid/appauth/built/authorization_service_configuration';
import * as appAuthFlags from '@openid/appauth/built/flags';
import { NodeBasedHandler, NodeCrypto } from '@openid/appauth/built/node_support';
import { RevokeTokenRequest } from '@openid/appauth/built/revoke_token_request';
import {
  GRANT_TYPE_AUTHORIZATION_CODE,
  GRANT_TYPE_REFRESH_TOKEN,
  TokenRequest,
} from '@openid/appauth/built/token_request';

import environment from './environment';
import storage from './storage';
import FitbitTokenRequestHandler from './FitbitTokenRequestHandler';

// TODO: PR against AppAuth to make this properly configurable
(appAuthFlags as any).IS_LOG = false;

const tokenHandler = new FitbitTokenRequestHandler();

function getAuthConfiguration() {
  const { apiUrl } = environment().config;
  return new AuthorizationServiceConfiguration({
    token_endpoint: `${apiUrl}/oauth2/token`,
    authorization_endpoint: `${apiUrl}/oauth2/authorize`,
    revocation_endpoint: `${apiUrl}/oauth2/revoke`,
  });
}

export async function refreshToken(refreshToken: string) {
  const { clientId } = environment().config;
  try {
    const response = await tokenHandler.performTokenRequest(
      getAuthConfiguration(),
      new TokenRequest({
        client_id: clientId,
        redirect_uri: '', // not needed
        grant_type: GRANT_TYPE_REFRESH_TOKEN,
        refresh_token: refreshToken,
      }),
    );
    await storage.set(response);
    return response.accessToken;
  } catch (ex) {
    await storage.clear();
    throw ex;
  }
}

function authError(msg: string) {
  return new Error(`Authorization error: ${msg}`);
}

function authorizationCallbackPromise(handler: AuthorizationRequestHandler) {
  const notifier = new AuthorizationNotifier();
  handler.setAuthorizationNotifier(notifier);

  return new Promise<{
    request: AuthorizationRequest,
    response: AuthorizationResponse,
  }>((resolve, reject) => {
    notifier.setAuthorizationListener((request, response, error) => {
      if (error) {
        if (error.errorDescription) {
          return reject(
            authError(`${error.error}: ${error.errorDescription}`),
          );
        }
        return reject(authError(error.error));
      }
      resolve({ request: request!, response: response! });
    });
  });
}

async function authorize() {
  const { clientId } = environment().config;

  const port = 13579;
  const redirectUri = `http://127.0.0.1:${port}`;

  const authorizationHandler = new NodeBasedHandler(port);
  authorizationHandler.performAuthorizationRequest(
    getAuthConfiguration(),
    new AuthorizationRequest(
      {
        client_id: clientId,
        redirect_uri: redirectUri,
        scope: 'profile',
        response_type: AuthorizationRequest.RESPONSE_TYPE_CODE,
      },
      new NodeCrypto(),
      true,
    ),
  );

  const { request, response } = await authorizationCallbackPromise(authorizationHandler);
  if (request.state !== response.state) throw authError('Mismatched state');
  return {
    redirectUri,
    code: response.code,
    pkceVerifier: request.internal!['code_verifier'],
  };
}

async function revoke(token: string) {
  await tokenHandler.performRevokeTokenRequest(
    getAuthConfiguration(),
    new RevokeTokenRequest({ token }),
  );
}

export async function login() {
  const { clientId } = environment().config;

  const { code, pkceVerifier, redirectUri } = await authorize();
  const response = await tokenHandler.performTokenRequest(
    getAuthConfiguration(),
    new TokenRequest({
      code,
      client_id: clientId,
      redirect_uri: redirectUri,
      grant_type: GRANT_TYPE_AUTHORIZATION_CODE,
      extras: { code_verifier: pkceVerifier },
    }),
  );
  await storage.set(response);
}

export async function logout() {
  const authData = await storage.get();
  if (!authData) return;
  await Promise.all([
    revoke(authData.accessToken),
    storage.clear(),
  ]);
}

export async function getAccessToken() {
  const authData = await storage.get();
  if (authData === null) return null;
  // Check for validity without any time buffer.
  if (authData.isValid(0)) return authData.accessToken;
  return refreshToken(authData.refreshToken!);
}
