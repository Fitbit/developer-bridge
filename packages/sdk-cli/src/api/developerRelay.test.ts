import nock from 'nock';
import stream from 'stream';
import websocketStream from 'websocket-stream';

import * as auth from '../auth';
import environment from '../auth/environment';
import * as developerRelay from './developerRelay';
import mockWithPromiseWaiter from '../testUtils/mockWithPromiseWaiter';

jest.mock('websocket-stream', () => jest.fn());
jest.mock('../auth');

const mockAppHost = {
  id: 'apphostA',
  displayName: 'App Host',
  roles: ['APP_HOST'],
  state: 'available',
};

const mockCompanionHost = {
  id: 'companionhostA',
  displayName: 'Companion Host',
  roles: ['COMPANION_HOST'],
  state: 'available',
};

let endpointMock: nock.Scope;

function relayHostToHost(relayHost: typeof mockAppHost) {
  return {
    displayName: relayHost.displayName,
    roles: relayHost.roles,
    available: relayHost.state === 'available',
    connect: expect.any(Function),
  };
}

function mockHostsResponse(status: number, payload?: any) {
  endpointMock
    .get('/1/user/-/developer-relay/hosts.json')
    .reply(status, payload);
}

function mockHostsSuccessResponse(hosts = [mockAppHost, mockCompanionHost]) {
  mockHostsResponse(200, { hosts });
}

function mockConnectionURLResponse(hostID: string, response: string) {
  endpointMock
    .post(`/1/user/-/developer-relay/hosts/${hostID}`)
    .reply(200, response, { 'Content-Type': 'text/uri-list' });
}

beforeEach(() => {
  (auth.getAccessToken as jest.Mock).mockResolvedValue('mockToken');
  endpointMock = nock(environment().config.apiUrl);
});

describe('hosts()', () => {
  it('returns list of connected hosts', async () => {
    mockHostsSuccessResponse();
    return expect(developerRelay.hosts()).resolves.toEqual([
      relayHostToHost(mockAppHost),
      relayHostToHost(mockCompanionHost),
    ]);
  });

  it('returns empty lists if no hosts are connected', async () => {
    mockHostsSuccessResponse([]);
    return expect(developerRelay.hosts()).resolves.toEqual([]);
  });

  it('parses a 403 response for error reasons', async () => {
    mockHostsResponse(403, {
      errors: [{ message: 'reason 1' }, { message: 'reason 2' }],
    });
    return expect(developerRelay.hosts()).rejects.toEqual(
      new Error('reason 1\nreason 2'),
    );
  });

  it('handles a 403 response with a malformed payload', async () => {
    mockHostsResponse(403, { message: 'reason 2' });
    return expect(developerRelay.hosts()).rejects.toMatchSnapshot();
  });

  it('handles a 403 response with a non JSON payload', async () => {
    mockHostsResponse(403);
    return expect(developerRelay.hosts()).rejects.toMatchSnapshot();
  });

  it('handles a non 403 error response', async () => {
    mockHostsResponse(500);
    return expect(developerRelay.hosts()).rejects.toMatchSnapshot();
  });
});

describe('connect()', () => {
  const mockConnectionURL = 'ws://device_url';
  let mockWebSocket: stream.Duplex;
  let socketPromise: Promise<void>;
  let connectPromise: Promise<stream.Duplex>;

  beforeEach(async () => {
    mockWebSocket = new stream.Duplex();
    mockHostsSuccessResponse([mockAppHost]);
    socketPromise = mockWithPromiseWaiter(
      websocketStream as any,
      mockWebSocket,
    );
    const host = (await developerRelay.hosts())[0];
    mockConnectionURLResponse(mockAppHost.id, `${mockConnectionURL}\r\n`);
    connectPromise = host.connect();
    await socketPromise;
  });

  describe('when the WebSocket connection is successful', () => {
    beforeEach(() => mockWebSocket.emit('connect'));
    afterEach(() => connectPromise);

    it('opens a websocket connection to the retrieved URL', () => {
      expect(websocketStream).toBeCalledWith(mockConnectionURL, {
        objectMode: true,
      });
    });

    it('returns a stream', () =>
      expect(connectPromise).resolves.toBe(mockWebSocket));
  });

  describe('when the WebSocket connection is unsucessful', () => {
    beforeEach(() =>
      mockWebSocket.emit('error', new Error('Connection failed!')),
    );

    it('throws', () =>
      expect(connectPromise).rejects.toThrowErrorMatchingSnapshot());
  });
});
