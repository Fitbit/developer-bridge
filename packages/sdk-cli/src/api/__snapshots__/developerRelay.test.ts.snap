// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect() when the WebSocket connection is unsucessful throws 1`] = `"Connection failed!"`;

exports[`hosts() handles a 403 response with a malformed payload 1`] = `[Error: Fetch of https://api.fitbit.com/1/user/-/developer-relay/hosts.json returns status 403 with payload: {"message":"reason 2"}]`;

exports[`hosts() handles a 403 response with a non JSON payload 1`] = `[Error: Fetch of https://api.fitbit.com/1/user/-/developer-relay/hosts.json returned status 403 Forbidden]`;

exports[`hosts() handles a non 403 error response 1`] = `[Error: Fetch of https://api.fitbit.com/1/user/-/developer-relay/hosts.json returned status 500 Internal Server Error]`;
